[5.1] Improvements to Arr::sortRecursive

mzur opened the pull request:
I've recently encountered a few issues with `Arr::sortRecursive` so I'd like to propose some improvements to this function. These issues were:

**Recursion stops if an array containing the key `0` is encountered:**

| Array | Expected result | Actual result |
| --- | --- | --- |
| `[['b' => 'x', 'a' => 'x']]` | `[['a' => 'x', 'b' => 'x']]` | `[['b' => 'x', 'a' => 'x']]` |

**Flat arrays were not sorted:**

| Array | Expected result | Actual result |
| --- | --- | --- |
| `['b', 'a']` | `['a', 'b']` | `['b', 'a']` |

To improve recognizing of "non-associative" arrays (having incrementing numerical keys starting with 0), I've added an `Arr::isAssoc` method.

The general behavior is now as follows:
Every "non-associative" array (like `[3, 2, 1]`) is sorted by values using `sort()`. Every "associative" array (like `['b' => 1, 'a' => 5]`) is sorted by keys using `ksort()`. The array is now passed by reference (to be consistent with functions like `sort` or `ksort`; also for the recursion to work properly) but the sorted array is returned, too (for compatibility reasons).

**Edit:** The test fails with PHP 7; I'll look into it.

GrahamCampbell commented:
Please don't make ANY changes to the phpdoc.

mzur commented:
The changes to the phpdoc were done by the PHP-CS-Fixer using the `.php_cs` of this repo.

GrahamCampbell commented:
Well, the changes are wrong, Please revert them.

mzur commented:
Sorry for that; I thought the CS-Fixer would know what to do.

I'm still looking into the PHP 7 issue. But are there any general opinions on this proposal?

GrahamCampbell commented on tests/Support/SupportArrTest.php:
```diff
@@ -0,0 +1,95 @@
+{
+    public function testIsAssoc()
+    {
+        $this->assertTrue(Arr::isAssoc([
```
Please put these arrays on one line in each case.

GrahamCampbell replied:
```php
$this->assertTrue(Arr::isAssoc(['a' => 'a', 0 => 'b']));
```

GrahamCampbell commented on src/Illuminate/Support/Arr.php:
```diff
@@ -273,6 +273,20 @@ public static function has($array, $key)
+    public static function isAssoc(array $array)
+    {
+        $keys = array_keys($array);
+        return array_keys($keys) !== $keys;
```
You must have a blank line before a return statement.

timgws replied:
@GrahamCampbell isn't there a blank link before the return statement?

GrahamCampbell commented on src/Illuminate/Support/Arr.php:
```diff
@@ -273,6 +273,20 @@ public static function has($array, $key)
+     * keys).
+     * 
+     * @param array $array
+     *
```
no newline here.

GrahamCampbell commented on src/Illuminate/Support/Arr.php:
```diff
@@ -273,6 +273,20 @@ public static function has($array, $key)
     }
 
     /**
+     * Determines if an array is associative (doesn't have sequential numeric 
```
The short description must not exceed one line.

GrahamCampbell commented on src/Illuminate/Support/Arr.php:
```diff
@@ -273,6 +273,20 @@ public static function has($array, $key)
+     * Determines if an array is associative (doesn't have sequential numeric 
+     * keys).
+     * 
+     * @param array $array
```
Please put 2 spaces either side of the type for param annotations.

lukasgeiter commented:
I think this makes sense.
Especially with the second example (flat array) I really would expect it to get sorted correctly.

mzur commented:
@GrahamCampbell Thanks for taking your time with this. I've just now learned that I had to explicitly select the `.php_cs` config file. I assumed that would be done automatically.

GrahamCampbell commented:
This doesn't work on php 7. You must make sure this works on php 7.

mzur commented:
Right. As I mentioned earlier, I'm on it but I haven't found a solution yet.

mzur committed:
Improves Arr::sortRecursive.
Recursion will now continue if a deeply nested non-assiciative array is encountered.
Also implements Arr::isAssoc to better distinguish between associative
and non-associuative arrays (having sequential numerical keys).

mzur commented:
Apparently the behavior of `sort()` changed between PHP 5.6 and 7 (also mentioned [here](https://github.com/bolt/bolt/issues/3275)).
Running the following code:
```php
$a = [
    [
        'mail' => 'joe@example.com',
        'name' => 'joe',
        'numbers' => [0, 1, 2],
    ],
    [
        'age' => 25,
        'mail' => 'jane@example.com',
        'name' => 'jane',
    ],
];
sort($a);
dd($a);
```
Yielded these results:

**PHP 5.6**
```php
array:2 [
  0 => array:3 [
    "mail" => "joe@example.com"
    "name" => "joe"
    "numbers" => array:3 [
      0 => 0
      1 => 1
      2 => 2
    ]
  ]
  1 => array:3 [
    "age" => 25
    "mail" => "jane@example.com"
    "name" => "jane"
  ]
]
```
**PHP 7**
```php
array:2 [
  0 => array:3 [
    "age" => 25
    "mail" => "jane@example.com"
    "name" => "jane"
  ]
  1 => array:3 [
    "mail" => "joe@example.com"
    "name" => "joe"
    "numbers" => array:3 [
      0 => 0
      1 => 1
      2 => 2
    ]
  ]
]
```
I was able to work around this issue by changing the length of one array.

taylorotwell commented:
Does this PR change the behavior of sortRecursive in ANY backwards in-compatibile way?

lukasgeiter commented:
@taylorotwell Yes it does. Even though it fixes the behavior, it is a BC break and should go into 5.2

mzur commented:
For the recursion to work properly, the argument passed to `sortRecursive` is now passed by reference. So it *may* have different behavior in cases like these:
```php
$sortedArray = array_sort_recursive($unsortedArray);
// do something with $unsortedArray and assume it wasn't changed
```
@lukasgeiter You're right, it probably should go into 5.2.
